sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -1237,20 +1250,79 @@ class Analyzer(
+      : (LogicalPlan, Seq[(Attribute, Attribute)]) = {
+      if (conflictPlanMap.contains(plan)) {
+        // If the plan is the one that conflict the with left one, we'd
+        // just replace it with the new plan and collect the rewrite
+        // attributes for the parent node.
+        val newRelation = conflictPlanMap(plan)
+        newRelation -> plan.output.zip(newRelation.output)
+      } else {
+        val attrMapping = new mutable.ArrayBuffer[(Attribute, Attribute)]()
+        val newPlan = plan.mapChildren { child =>
+          // If not, we'd rewrite child plan recursively until we find the
+          // conflict node or reach the leaf node.
+          val (newChild, childAttrMapping) = rewritePlan(child, conflictPlanMap)
+          // Only return rewrite attributes which could be used by the parent node.
+          // Otherwise, it could introduce duplicate rewrite attributes. For example,
+          // for the following plan, if we don't do filter for the `childAttrMapping`,
+          // the node `SubqueryAlias b` will return rewrite attribute of [kind#220 -> kind#228]
+          // (which is from the conflict plan `Project [id#218, foo AS kind#228]`), and the node
+          // `SubqueryAlias c` will return rewrite attribute of [kind#220 -> kind#229] (which
+          // is from the conflict plan `Project [id#227, foo AS kind#229]`). As a result, the top
+          // Join will have duplicated rewrite attribute.
+          //
+          // The problem is, the plan `Join Inner, (kind#229 = kind#223)` shouldn't keep returning
+          // rewrite attribute of [kind#220 -> kind#229] to its parent node `Project [id#227]` as
+          // it doesn't really need it.
+          //
+          // Join Inner, (id#218 = id#227)
+          // :- SubqueryAlias b
+          // :  +- Project [id#218, foo AS kind#228]
+          // :     +- SubqueryAlias a
+          // :        +- Project [1 AS id#218]
+          // :           +- OneRowRelation
+          // +- SubqueryAlias c
+          //    +- Project [id#227]
+          //       +- Join Inner, (kind#229 = kind#223)
+          //          :- SubqueryAlias l
+          //          :  +- SubqueryAlias b
+          //          :     +- Project [id#227, foo AS kind#229]
+          //          :        +- SubqueryAlias a
+          //          :           +- Project [1 AS id#227]
+          //          :              +- OneRowRelation
+          //          +- SubqueryAlias r
+          //             +- SubqueryAlias b
+          //                +- Project [id#224, foo AS kind#223]
+          //                   +- SubqueryAlias a
+          //                      +- Project [1 AS id#224]
+          //                         +- OneRowRelation
+          attrMapping ++= childAttrMapping.filter { case (oldAttr, _) =>
+            (plan.references ++ plan.outputSet ++ plan.producedAttributes).contains(oldAttr)
```

Comment:
Hi all @cloud-fan @maropu @viirya , please take a look at the latest change according to SPARK-32280 meitoned by @imback82 . I've also added a new test migrated from SPARK-32280.

Reply:
Can you add [SPARK-32280] in the PR title too?

Reply:
why do we need `plan.producedAttributes`?

Reply:
and seems we only need `plan.outputSet`: if an attribute is not output by the current plan, it's useless to propagate this attribute to the parent nodes.

Reply:
The above is probably good enough as the comment here.

Reply:
Seems we need `plan.outputSet` too. I think it is similar to case of aliases in projection above you added.

Reply:
`plan.references` is also necessary? For example,  Join's condition references the attributes but its output doesn't contain the attribute.

Reply:
> Seems we need plan.outputSet too. I think it is similar to case of aliases in projection above you added.

Yeah, "aliases in projection" is another case. And I assume you were referring `plan.references` too. @viirya

Reply:
`plan.references` won't be the output, right? I guess they won't conflict with others?

Reply:
Yes, they won't. I mean, you probably have a typo above:

 > Seems we need plan.outputSet too.

IIUC, it sounds like you're actually meaning the ` plan.references` according to the context.

Reply:
Nvm, I actually referred `plan.outputSet`. But now I see `plan.references` is also needed.